src/io/cmap.rs

```diff
@@ -0,0 +1,1038 @@
+//! Structs amd trait to read files in CMAP format.
+//!
+//! Note: This module only supports version v0.1 of the CMAP format for now.
+//!
+//! # Example
+//!
+//! ## Reader
+//!
+//! In this example, we parse a CMAP file, iterate over its records. and output some statistics.
+//!
+//! ```rust
+//! use bio::io::cmap;
+//!
+//! let path = "tests/resources/valid_input.cmap";
+//! let mut reader = cmap::Reader::from_path(&path).unwrap();
+//!
+//! let mut num_contigs = 0;
+//! let mut num_labels = 0;
+//!
+//! while let Some(Ok(mut record)) = reader.next() {
+//!     num_contigs += 1;
+//!     num_labels += record.labels().len() - 1;
+//! }
+//!
+//! println!("Number of contigs: {}", num_contigs);
+//! println!("Number of labels: {}", num_labels);
+//! ```
+//!
+//! ## Container
+//!
+//! If feasible, we can also build a container over the content of a CMAP file.
+//!
+//! ```rust
+//! use bio::io::cmap;
+//!
+//! let path = "tests/resources/valid_input.cmap";
+//! let mut container = cmap::Container::from_path(&path).unwrap();
+//!
+//! println!("CMAP header: {:?}", container.header());
+//! println!("CMAP label: {:?}", container.labels());
+//! ```
+
+use anyhow::{bail, Context, Result};
+use derive_new::new;
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+
+use crate::io::om_utils::Error;
+use crate::io::om_utils::NextToErr;
+use crate::io::om_utils::ParseToErr;
+
+/// A label as given in a CMAP file.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Label {
+    id: u32,
+    pos: f64,
+    std_dev: f32,
+    coverage: u32,
+    occurrence: u32,
+}
+
+/// A CMAP record.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Record {
+    id: u32,
+    len: f64,
+    num_labels: u32,
+    label_channel: u8,
+    labels: Vec<Label>,
+}
```

Comment:
Why does a cmap record not have a contig name?

Reply:
There are no contig names in CMAPs; they are identified solely by their integer ID.

Reply:
I see. Where are the contig names stored then?

Reply:
Then the .fai of the fasta is the right source.